File: robot/PREN11-PI/PiController/ServerCommunicator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using SocketIOClient;

namespace PiController
{
    public class ServerCommunicator
    {
        SocketIO client;
        public ServerCommunicator(string url)
        {
            var initTask = init(url);
            initTask.Wait();
        }

        private async Task init(string url)
        {
            Console.WriteLine("Setting up SocketIO Client");
            client = new SocketIO(url);
            client.OnConnected += async (sender, e) =>
            {
                Console.WriteLine("Connected To Server");
            };
            await client.ConnectAsync();
        }

        public void SendStart()
        {
            SendEvent("START");
        }

        public void SendEnd()
        {
            SendEvent("END");
        }

        private void SendEvent(string sendEvent)
        {
            Console.WriteLine("Sending " + sendEvent + " Event to Server");
            client.EmitAsync("RobotInput", response =>
                {
                // You can print the returned data first to decide what to do next.
                // output: [{"result":true,"message":"Prometheus - server"}]
                Console.WriteLine(response);
                }
                    , "{\"event_type\":\"" + sendEvent + "\"}").Wait();
        }

        public void SendPlant(string plantName)
        {
            Console.WriteLine("Sending Plant" + plantName + " to Server");
            client.EmitAsync("RobotInput", response =>
            {
                // You can print the returned data first to decide what to do next.
                // output: [{"result":true,"message":"Prometheus - server"}]
                Console.WriteLine(response);
            }
                    , "{\"event_type\":\"PLANT\", \"event_value\":{\"plantName\":\""+plantName+"\"}}").Wait();
        }
    }
}

File: robot/pren-java/src/main/java/com/pren11/server/ServerCommunicator.java
package com.pren11.server;
import com.pren11.Config;
import io.socket.client.IO;
import io.socket.client.Socket;
import io.socket.emitter.Emitter;

import java.net.URI;

import static java.lang.System.out;
import static java.lang.System.err;

public class ServerCommunicator {

    Socket io;

    public ServerCommunicator(String url){
        init(url);
    }

    private void init(String url)
    {
        if(Config.OFFLINE_MODE) return;
        out.println("Setting up SocketIO Client");
        URI uri = URI.create(url);
        IO.Options options = IO.Options.builder()
                // ...
                .build();

        io = IO.socket(uri, options);
        io.on("connection", objects -> {
            out.println("Socket Setup with Server");
        });

        io.on("error", o -> {
            out.println("ERRROR SOCKET: " + o.toString());
        });

        io.connect();

        do{
            System.out.println("Is Connected? " + io.connected());
        }while(io.connected() == false);
        System.out.println("Final Connection Test? " + io.connected());


       // await client.ConnectAsync();
        Runtime.getRuntime().addShutdownHook(new Thread(new Runnable() {

            public void run() {
                // Do what you want when the application is stopping
                System.out.println("OHNOOOOOO, gotta tip quick");
                disconnect();
            }
        }));
    }

    public void disconnect() {
        io.disconnect();
        io.close();
    }


    public void sendStart()
    {
        if(Config.OFFLINE_MODE) return;
        sendEvent("START");
    }

    public void sendEnd()
    {
        if(Config.OFFLINE_MODE) return;
        sendEvent("END");
    }

    private void sendEvent(String sendEvent)
    {
        if(Config.OFFLINE_MODE) return;
        out.println("Sending " + sendEvent + " Event to Server, active: " + io.isActive() + " connected: " + io.connected());
        io.emit("RobotInput", "{\"event_type\":\"" + sendEvent + "\"}");

    }

    public void sendPlant(String plantName, String image)
    {
        if(Config.OFFLINE_MODE) return;
        out.println("Sending Plant" + plantName + " to Server");
        var emitter = io.emit("RobotInput","{\"event_type\":\"PLANT\", \"event_value\":{\"plantName\":\""+plantName+"\",\"image\":\""+image+"\"}}");
    }


}
